### PR TITLE
feature: knn kd tree fix for parallel

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
## Description

_Add a comprehensive description of proposed changes_

_List associated issue number(s) if exist(s): #6 (for example)_

_Documentation PR (if needed): #1340 (for example)_

_Benchmarks PR (if needed): https://github.com/IntelPython/scikit-learn_bench/pull/155 (for example)_

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.
- [ ] I have provided justification why quality metrics have changed or why changes are not expected.
- [ ] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i [hunk 1]
```diff
@@ -61,39 +61,70 @@ using namespace kdtree_knn_classification::internal;
...
     DAAL_FORCEINLINE void push(const T & value)
     {
-        _data[_last = (_last + 1) & _sizeMinus1] = value;
+        if (_count >= _capacity) // Check if capacity is exceeded
+        {
+            services::Status status = grow(); // Grow if necessary
+            //DAAL_CHECK_STATUS_VAR(status);
+        }
+
+        _data[_last = (_last + 1) & _sizeMinus1] = value; // Add element to queue
```
Comment (avolkov-intel) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:115]: It seems that for this condition to make sense Size should be the power of 2. Since this can depend on value provided by user it is better to use (_last + 1) % size

Comment (avolkov-intel) [reply cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:115]: Or explicitly force size to be the power of two

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i [hunk 2]
```diff
@@ -189,9 +238,7 @@ Status KNNClassificationTrainBatchKernel<algorithmFpType, training::defaultDense
     typedef BoundingBox<algorithmFpType> BBox;
 
     const algorithmFpType base = 2.0;
-    // The queue size is not impacted by number of threads.
-    // All operations with the queue are done not in the threader_for primitives.
-    const size_t queueSize = 2 * Math::sPowx(base, Math::sCeil(Math::sLog(__KDTREE_FIRST_PART_LEAF_NODES_PER_THREAD) / Math::sLog(base)));
+    const size_t queueSize     = 2 * Math::sPowx(base, Math::sCeil(Math::sLog(__KDTREE_FIRST_PART_LEAF_NODES_PER_THREAD) / Math::sLog(base)));
```
Comment (avolkov-intel) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:241]: Add comment about the meaning of that value: (seems it is the lowest power of two that is greater than log_2(__KDTREE_FIRST_PART_LEAF_NODES_PER_THREAD) multiplied by 2)

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i [hunk 3]
```diff
@@ -61,39 +61,63 @@ using namespace kdtree_knn_classification::internal;
...
+            grow();
+        }
+
+        _data[_last] = value;
+        ++_last;
         ++_count;
     }
 
-    DAAL_FORCEINLINE T pop()
+    T pop()
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:117]: Looks like `--_last;` is missing.
Maybe that's the reason why `grow()` started to be required?

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i [hunk 4]
```diff
@@ -61,39 +61,63 @@ using namespace kdtree_knn_classification::internal;
 template <typename T, CpuType cpu>
 class Queue
 {
+    static const size_t defaultSize = 4;
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:64]: Please add comments to the magic constants.

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i (hunk 3 above)
Comment (Alexandr-Solovev) [reply cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:117]: Is it necessary to call `--last`. I move the index of the first element, but I shouldnt touch the last one. _last indicates the next available position for insertion, not the index of the last element

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i (hunk 1 above)
Comment (Alexandr-Solovev) [reply cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i]: the queue class has been updated

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i [hunk 5]
```diff
@@ -103,12 +130,28 @@ public:
     size_t size() const { return _count; }
 
 private:
+    void grow()
+    {
+        size_t newCapacity = (_capacity == 0 ? defaultSize : _capacity * 2);
+        T * newData        = static_cast<T *>(service_malloc<T, cpu>(newCapacity));
+
+        for (size_t i = 0; i < _count; ++i)
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:138]: Please use `daal_memcpy_s` if the typical size of the loop is big enough. I.e. 20+ iterations.

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i [hunk 6]
```diff
@@ -1202,79 +1235,69 @@ Status KNNClassificationTrainBatchKernel<algorithmFpType, training::defaultDense
...
-            localTLS.reduce([=, &isNeedToReindex](Local * ptr) -> void {
-                if (ptr && ptr->extraKDTreeNodes)
+            int result = 0;
+
+            size_t actualNodeCount = lastNodeIndex;
+            localTLS.reduce([=, &actualNodeCount](Local * ptr) -> void {
+                if (ptr)
                 {
-                    isNeedToReindex = true;
+                    actualNodeCount += ptr->nodeIndex - firstNodeIndex[ptr->threadIndex];
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i:1244]: Would be interesting to see how this change impacts the performance.
But anyway, the change is justified as it fixes the crashes.

**Review by Alexsandruss — CHANGES_REQUESTED** [review]
Next example gives seg. fault with this PR:
```python
from sklearnex.neighbors import KNeighborsClassifier
from sklearn.datasets import make_classification
import logging as lg


lg.getLogger().setLevel("DEBUG")

x, y = make_classification(
    n_samples=200000,
    n_features=8,
    random_state=42,
)
knn = KNeighborsClassifier(algorithm="kd_tree")
knn.fit(x, y)
print("Score:", knn.score(x, y))
```

Backtrace:

> daal::algorithms::kdtree_knn_classification::prediction::internal::KNNClassificationPredictKernel<double, (daal::algorithms::kdtree_knn_classification::prediction::Method)0, (daal::CpuType)6>::findNearestNeighbors(double const*, daal::internal::Heap<daal::internal::GlobalNeighbors<double, (daal::CpuType)6>, (daal::CpuType)6>&, daal::algorithms::kdtree_knn_classification::internal::Stack<daal::algorithms::kdtree_knn_classification::prediction::internal::SearchNode<double>, (daal::CpuType)6>&, unsigned long, double, daal::algorithms::kdtree_knn_classification::interface1::KDTreeTable const&, unsigned long, daal::data_management::interface1::NumericTable const&, bool, daal::services::internal::DynamicArray<double*, daal::services::internal::ScalableMalloc<double*, (daal::CpuType)6>, daal::services::internal::DefaultConstructionPolicy<double*, (daal::CpuType)6, true>, (daal::CpuType)6>&)
